Course Project

kitzsable opened the pull request:
Course Project

Commit:
Course Project

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/model/Session.java:
```diff
@@ -0,0 +1,24 @@
+    private String name;
+
+    @Column
+    private Date date;
```
Лучше использовать LocalDate или LocalDateTime

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/repository/SelectedAnswerRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+@Repository
+public interface SelectedAnswerRepository extends CrudRepository<SelectedAnswer, Long> {
+    List<SelectedAnswer> findAllBySession_Name(String name);
```
Не стоит использовать подчеркивания в именах методов. Нет способов реализовать метод иначе?

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+
+    // Идентификаторы журналов
+    public static final String QUESTIONS_JOURNAL_ID = "questions";
+    public static final String SESSIONS_JOURNAL_ID = "sessions";
```
В качестве улучшения, можно подумать в сторону создания отдельного Enum для кодов журналов. Исправление, на Ваше усмотрение.

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * Метод сохранения экземпляра журнала
+     * @param journal Журнал
+     */
```
Javadoc это хорошо! :)

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+     */
+    @Override
+    public JournalLineResponseDTO getJournalLineResponse(String id, JournalLineRequestDTO request) {
+        // Ответ со строками журнала и массив строк журнала
```
Странный комментарий, наверно имелось ввиду общее количество строк и строки текущей страницы?

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+                        question ->
+                                new QuestionDTO(question,
+                                        answerRepository.findAllByQuestion(question)),
+                        request.filters.isEmpty() || request.filters.get(0).value==null ?
```
Есть кейсы когда isEmpty() недостаточно? Не совсем понятно, зачем дополнительно проверять, что нулевой элемент null

kitzsable replied:
Да, если на клиенте сначала отфильтроввать вопросу по какому то количеству ответов, а потом попробовать снова получить все строки, то в запросе список фильтров не пустой, а с одним элементом, в котором значение null.

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+                                questionDTO -> true :
+                                questionDTO ->
+                                        questionDTO.answers.size() ==
+                                                Double.parseDouble(request.filters.get(0).value));
```
Слишком большой кусок логики засунут на вход метода getJournalLines(). Читать очень тяжело. Если появилась необходимость в таких больших кусках кода, то надо выносить в отдельные методы.

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+                response = new JournalLineResponseDTO( (int)sessionRepository.count(), lines);
+                break;
+            default:
+                throw new RuntimeException("В url содержится неизвестный id журнала: " + id);
```
Пользователь ничего не сможет сделать с URL, а ошибка именно ему прилетит. Я бы просто оставил текст, что не найден журнал с кодом таким-то.

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+                            sessionRepository::findByNameContainingIgnoreCase,
+                            SessionDTO::new,
+                            sessionDTO -> true);
+                response = new JournalLineResponseDTO( (int)sessionRepository.count(), lines);
```
Зачем здесь кастовать к int? Может быть где то поменять ожидаемые типы данных?

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/JournalServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+        // элементы преобразуются к соответствующим DTO,
+        // проводится фильтрация по заданному предикату,
+        // возвращается получившийся список из DTO строк журнала
+        return generator.apply(request.search, pageRequest)
```
Я думаю что здесь комментарий не нужен. Код достаточно читабельный и у читающего не должно возникнуть вопросов, что делает этот код. Обычно комментарии пишутся когда есть не очевидные нюансы

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/QuestionServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        // Удаление вопроса
+        questionRepository.deleteById(Long.parseLong(questionDTO.id));
+        // Сохранение нового вопроса и возвращение DTO сохраненнго вопроса
+        return save(questionDTO);
```
Слишком много комментариев в методе. Это не нужно и только ухудшает читаемость

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/QuestionServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        // Удаление ответов на полученный вопрос из БД
+        answerRepository.deleteAllByQuestion(question);
+        // Удаление вопроса
+        questionRepository.deleteById(Long.parseLong(questionDTO.id));
```
Зачем удалять вопрос, когда мы хотим его обновить? Это неправильно

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,198 @@
+        // затем суммируются в общее колличество баллов,
+        // а после считается процент набранных балло от возможного максимума
+        // и округляется до целого значения
+        int percent = (int) Math.round(
```
Дробного процента быть не может?

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,198 @@
+        // Если количество вопросов не больше 5, то
+        // возвращается список вопросов преобразованных в DTO.
+        // Иначе генерируются номера, вопросы под которыми
+        // преобразуются в DTO и сохраняются в виде списка.
```
разве сохраняются?

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,198 @@
+                                    answerRepository.findAllByQuestion(question)))
+                    .collect(Collectors.toList());
+        } else {
+            result = generateIndexes(questions.size())
```
Рандомные вопросы из списка можно было получить проще, например использовав Collections.shuffle, перемешать список и взять из него первые N элементов.
Другой вариант использовать в качестве результирующей переменно - Set вместо List, значения в Set уникальны. Тогда можно было бы просто брать значения из массив вопросов по рандомному индексу и не следить за ним.

Mrak2017 commented on unit-11/app-server/src/main/java/ru/kitzsable/appserver/transfer/AnswerDTO.java:
```diff
@@ -0,0 +1,20 @@
+    public AnswerDTO(Answer answer) {
+        id = answer.getId().toString();
+        answerText = answer.getAnswerText();
+        isCorrect = null;
```
При редактировании вопроса и его списка ответов на клиенте, мы увидим какие ответы были ранее отмечены как правильные?

Mrak2017 commented on unit-11/app-server/src/test/java/ru/kitzsable/appserver/AppServerApplicationTests.java:
```diff
@@ -0,0 +1,475 @@
+				.defaultPageSize(5L)
+				.build();
+		journalService.save(journal);
+		assertEquals(journalService.getJournal("testJournal"), new JournalResponseDTO(journal));
```
Лучше сравнивать айдишники, а не сами объекты

Mrak2017 commented on unit-11/app-server/src/test/java/ru/kitzsable/appserver/AppServerApplicationTests.java:
```diff
@@ -0,0 +1,475 @@
+				SessionQuestionAnswerDTO questionAnswer = new SessionQuestionAnswerDTO();
+				questionAnswer.id = String.valueOf(answer.getId());
+				questionAnswer.isSelected = isFirst[0];
+				isFirst[0] =false;
```
Какая то слишком навороченная логика с isFirst. Лучше явно простым способом перечислить все входные данные для последующих проверок.

Mrak2017 reviewed (changes requested):
В целом работа хорошая, но хочется более прозрачных тестов (1-2 шт) на корректность расчета финального процента по сессии.
Читабельность кода является важным пунктом.

Commit:
1. Вместо Date использован LocalDate
2. Избавился от метода с подчеркиванием
3. В методе получения строк с вопросами вынесена в отдельный метод логика создания фильтра
4. Изменен метод обновения вопроса (избавился от удалений)
5. Добавил возможжность дробного процента
6. Исправил сокрытие верных ответов от пользователя
7. Немного поменял тест на расчет процента, добавил коментариев и assert-ов

Mrak2017 commented:
Окей, правки принимаются.
Спасибо за исправления
